include/matx/operators/interp.h

```diff
@@ -0,0 +1,474 @@
+          MATX_ASSERT(err == cudaSuccess, matxCudaError);
+
+          if constexpr (std::is_same_v<value_type, float>) {
+            cusparse_status = cusparseSgtsv2(
+              handle,       // cuSPARSE handle
+              n,            // Size of the system
+              1,            // Batch count
+              ptr_dl_,      // Sub-diagonal
+              ptr_d_,       // Main diagonal
+              ptr_du_,      // Super-diagonal
+              ptr_m_,       // Right-hand side and solution
+              n,            // batch_stride
+              workspace);   // Workspace buffer
+          } else if constexpr (std::is_same_v<value_type, double>) {
+            cusparse_status = cusparseDgtsv2(
+              handle,       // cuSPARSE handle
+              n,            // Size of the system
+              1,            // Batch count
+              ptr_dl_,      // Sub-diagonal
+              ptr_d_,       // Main diagonal
+              ptr_du_,      // Super-diagonal
+              ptr_m_,       // Right-hand side and solution
+              n,            // batch_stride
+              workspace);   // Workspace buffer
+          }
+          MATX_ASSERT(cusparse_status == CUSPARSE_STATUS_SUCCESS, matxCudaError);
+          // cleanup
+          err = cudaFreeAsync(workspace, stream);
+          MATX_ASSERT(err == cudaSuccess, matxCudaError); 
+          cusparse_status = cusparseDestroy(handle);
+          MATX_ASSERT(cusparse_status == CUSPARSE_STATUS_SUCCESS, matxCudaError);
+          matxFree(ptr_d_);
+          matxFree(ptr_dl_);
+          matxFree(ptr_du_);
+        }
+
+      }
+
+      template <typename ShapeType, typename Executor>
+      __MATX_INLINE__ void PostRun([[maybe_unused]] ShapeType &&shape, 
+                                  [[maybe_unused]] Executor &&ex) const noexcept {
+        if constexpr (std::is_same_v<Method, InterpMethodSpline>) {
+          matxFree(ptr_m_);
+        }
+      }
+
+
+      template <typename... Is>
+      __MATX_INLINE__ __MATX_DEVICE__ __MATX_HOST__ decltype(auto) operator()(Is... indices) const
+      {
+        auto x_query = xq_(indices...);
+        auto [idx_low, idx_high] = searchsorted(x_query);
+
+        return interpolate(x_query, idx_low, idx_high);
+      }
+
+    };
+  } // namespace detail
+
+
+/**
+ * Interpolation operator with specified method
+ *
+ * @tparam OpX
+ *   Type of sample points vector
+ * @tparam OpV
+ *   Type of sample values vector
+ * @tparam OpXQ
+ *   Type of query points
+ * @tparam Method
+ *   Interpolation method type (InterpMethodLinear, InterpMethodNearest, InterpMethodNext, InterpMethodPrev)
+ * @param x
+ *   Sample points (must be sorted in ascending order)
+ * @param v
+ *   Sample values (must be the same size as x)
+ * @param xq
+ *   Query points where to interpolate
+ * @returns Operator that interpolates values at query points
+ */
+template <typename Method = InterpMethodLinear, typename OpX, typename OpV, typename OpXQ>
```

Comment:
I think we should try to avoid the method as a template parameter in the user-facing code here. Since this is a fairly complex function anyways, I think an extra branch in `operator()` to check the method is going to be insignificant in the runtime. You can keep roughly what you have already as separate functions and just use overloads to catch the right one rather than sfinae

Reply:
Are you suggesting using an enum instead of the template parameters?

One downside of that approach is that we can't support options (e.g. for boundary conditions or extrapolation)

Reply:
Yes. In general we've moved away from template parameters as much as possible due to trying to match python and matlab as much as possible. I think this is one of those cases where there's enough work to do inside each branch that the branch itself would be insignificant. I was just suggesting that you could keep it as templatized internally, but you would have a single `interpolate` method that would do something like: 
```
if (method == method_a)
  interpolate<METHOD_A, ...>();
else if (method == method_b)

etc
```

Reply:
Okay, I can make that change.

Since it will typically be a compile-time constant, I imagine the branches should be optimized away.